Develop stream 2025-01-20 (#455)
ROCm/hipCUB: CHANGELOG.md

  * `ScanTileState`
  * `ReduceByKeyScanTileState`
  * `TilePrefixCallbackOp`

> * Added overload of `BlockScan::InclusiveScan` accepting an initial value that seeds the scan.

Reviewer: ```suggestion
* Added an overload of `BlockScan::InclusiveScan` that accepts an initial value to seed the scan.
```

Author: Changed it in the newest commit.

> * Added overload of `WarpScan::InclusiveScan` accepting an initial value that seeds the scan.

Reviewer: ```suggestion
* Added an overload of `WarpScan::InclusiveScan` that accepts an initial value to seed the scan.
```

Author: Changed it in the newest commit.

> * Added `UnrolledThreadLoad` and `UnrolledCopy` for CUB parity which do a multiple thread loads or a copy instructions.

Reviewer: ```suggestion
* `UnrolledThreadLoad`, `UnrolledCopy`, and `ThreadLoadVolatilePointer` were added to align hipCUB with CUB. ```

I understood "CUB parity" to mean that the functions were added because they were in CUB but not hipCUB. If this isn't the case, let me know.

Author: Changed it in the newest commit.

> * Added `ThreadLoadVolatilePointer` for CUB parity.

Reviewer: Combined this with the previous item. The line can be deleted.

Author: Changed it in the newest commit.

> * Added `IterateThreadStore` struct to execute multiple stores for cuda parity.

Reviewer: ```suggestion
* `ThreadStoreVolatilePtr` and the `IterateThreadStore` struct were added to align with CUDA.
```

If this isn't what was meant by "parity" let me know.

Author: Changed it in the newest commit.

> * Added `ThreadStoreVolatilePtr` for cuda parity.

Reviewer: Combined with the previous item; can be deleted.

Author: Changed it in the newest commit.

### Changed

> * The NVIDIA backend now requires CUB, Thrust and libcu++ 2.6.0. If it is not found it will be downloaded from the NVIDIA CCCL repository.

Reviewer: ```suggestion
* The NVIDIA backend now requires CUB, Thrust, and libcu++ 2.6.0. If they aren't found, they will be downloaded from the NVIDIA CCCL repository.
```

Author: Changed it in the newest commit.

> * Updated `thread_load` and `thread_store` API for cuda parity.

Reviewer: ```suggestion
* Updated `thread_load` and `thread_store` to align with CUDA.
```

If this isn't what was meant by "parity"

Author: Changed it in the newest commit.

> * Changed symbol visibility: all kernels now have hidden visibility, inline namespace with the library version is added, e.g. `hipcub::HIPCUB_300400_NS::...`. This allows to link multiple libraries built with different versions of hipCUB.

Reviewer: ```suggestion
* All kernels now have hidden visibility with inline namespaces that include the library version, letting the user link multiple libraries built with different versions of hipCUB.
```

I think this is what was meant. If you want an example added in, it would need to be a "before and after" or a very qualified one so that the user can see the change.

Author: The hidden visibility is something separate from the namespace change. Is something like this clear? 
```suggestion
* All kernels now have hidden symbol visibility. All symbols now have inline namespaces that include the library version, (for example, hipcub::HIPCUB_300400_NS::symbol instead of hipcub::symbol), letting the user link multiple libraries built with different versions of hipCUB.
```

Reviewer: Yup! That looks good!

### Known issues

> * The deprecated `BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails` are not present in CCCL (CUB) from release 2.6.0 on, and thus neither they are in hipCUB's CUB backend for versions strictly greater than 3.4.0. These will be definitely removed from both backends in a future release.

Reviewer: **This should be under "Removed" rather than Known Issues**

```suggestion
*`BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails`, and `BlockAdjacentDifference::FlagHeadsAndTails` have been deprecated. They were removed in CCCL (CUB) 2.6.0 and have been removed in all versions of hipCUB's CUB backend later than 3.4.0. 
```

When you say "strictly greater than 3.4.0" do you mean from 3.4.1 and on, or do you mean something like 3.4.0.1 and on?

Also I'm confused by this line: "These will be definitely removed from both backends in a future release."

Have they been removed or not? Or do you mean they'll be removed from a different backend in a future release? If they will be removed from a different backend in a future release, that info should be added to the "Upcoming Changes" section.

Reviewer 2: > When you say "strictly greater than 3.4.0" do you mean from 3.4.1 and on, or do you mean something like 3.4.0.1 and on?

AFAIK we only do `mayor.minor.patch` kind of releases, hence it's meant to be from 3.4.1 on. However, even if a version such as 3.4.0.1 is possible, the update to CCCL 2.6.0 has been done after the 3.4.0 release of hipCUB, so regardless of the next version number the dependency on CCCL 2.6.0 will be there and hence the absence of these functions in the CUB backend of hipCUB.

> Also I'm confused by this line: "These will be definitely removed from both backends in a future release."
>
> Have they been removed or not? Or do you mean they'll be removed from a different backend in a future release?

They have been removed from CCCL/CUB (and thus from the CUB backend of hipCUB) as mentioned. However, as discussed internally, these should not be removed from the rocPRIM backend until the next ROCm mayor release so what the sentence means is that these will be dropped from the rocPRIM backend as well in some future release when the next ROCm mayor version is released.

> If they will be removed from a different backend in a future release, that info should be added to the "Upcoming Changes" section.

I see, we'll do this!

Reviewer 2: Oh and also, regarding the suggested changes:
> *`BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails`, and `BlockAdjacentDifference::FlagHeadsAndTails` have been deprecated.

These were already deprecated, so I think the original sentence should be ok. Perhaps something like:

```markdown
* The already-deprecated `BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails` ...
```

would make it more clear?

Reviewer: In which version were they deprecated? That could make it clearer. Something like "[functions] that were deprecated in [version] have now been removed"

Reviewer: How does this sound (with the right version put in instead of VERSION_X obvs :) ):

```
* `BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails` have been removed from hipCUB's CUB backend as of version 3.4.1. They were already removed from CCCL (CUB) as of version 2.6.0 and deprecated in VERSION_X of hipCUB's CUB backend.
```

Reviewer 2: That sounds good! My final suggestion after some re-writting to clarify which version is hipCUB's and which is CCCL's:
```markdown
### Known issues

* `BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails` have been removed from hipCUB's CUB backend. They were already deprecated as of version 2.12.0 of hipCUB and they were removed from CCCL (CUB) as of CCCL's 2.6.0 release.
...

### Upcoming Changes

* `BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails`, already deprecated as of version 2.12.0 of hipCUB, will be removed from the rocPRIM backend in a future release for the next ROCm major version (ROCm 7.0.0).

```

Reviewer: I'd smooth it a bit, but otherwise I'm good with this! Clarity is very appreciated! :)

`BlockAdjacentDifference::FlagHeads`, `BlockAdjacentDifference::FlagTails` and `BlockAdjacentDifference::FlagHeadsAndTails` were deprecated as of version 2.12.0 of hipCUB, and will be removed from the rocPRIM backend in a future release for the next ROCm major version (ROCm 7.0.0).

Reviewer: I'm gonna approve so you don't have to wait for me.

Reviewer 2: done!
